### PR TITLE
Nest auth

### PR DESCRIPTION
**Pull request by YuraOkolodko**

### CODE REVIEW

File: app/.env.example [hunk 1]
```diff
@@ -4,3 +4,5 @@ PORT=5432
 DB_NAME=dev
 USERNAME=postgres
 PASSWORD=example
+SECRET=secret
```
Comment (erikash) [inline app/.env.example:7]: rename to JW_SECRET

File: app/src/app.module.ts [hunk 2]
```diff
@@ -1,15 +1,25 @@
...
+      isGlobal: true,
     }),
+    UsersModule,
+    AuthModule,
+    CustomersModule,
   ],
-  controllers: [AppController],
-  providers: [AppService],
+  controllers: [AppController, AuthController, CustomersController],
+  providers: [AppService, CustomersService],
```
Comment (erikash) [inline app/src/app.module.ts:23]: You should probably register `CustomersService` as a provider in the `CustomersModule`

File: app/src/auth/auth.module.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+@Module({
+  imports: [
+      UsersModule,
+      PassportModule,
+      JwtModule.registerAsync({
+          inject: [ConfigService],
+          useFactory: async (configService: ConfigService) => ({
+              secret: configService.get<string>('SECRET'),
+              signOptions: { expiresIn: '60s' },
```
Comment (erikash) [inline app/src/auth/auth.module.ts:18]: 60 seconds really is the best practice?

File: app/src/auth/auth.module.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+          }),
+      }),
+  ],
+  providers: [
+      AuthService,
+      LocalStrategy,
+      JwtStrategy,
+      ConfigService
```
Comment (erikash) [inline app/src/auth/auth.module.ts:26]: You probably don't need to register `ConfigService` as a provider, you can import `ConfigModule` instead.

File: app/src/auth/auth.service.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    async validateUser(username: string, pass: string): Promise<any> {
+        const user = await this.usersService.findOne(username);
+        if (user && user.password === pass) {
+            const { password, ...result } = user;
+            return result;
+        }
+        return null;
+    }
+
+    async login(user: any) {
```
Comment (erikash) [inline app/src/auth/auth.service.ts:21]: Why `any`?

File: app/src/auth/strategies/headerapikey.strategy.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    ) {
+        super({ header: 'X-API-KEY', prefix: '' },
+            true,
+            async (apiKey: string, done: any) => {
+                return this.validate(apiKey, done);
+            });
+    }
+
+    public validate = (apiKey: string, done: (error: Error | null, data: any) => {}) => {
+        if (this.configService.get<string>('API_KEY') === apiKey) {
```
Comment (erikash) [inline app/src/auth/strategies/headerapikey.strategy.ts:19]: Could you promisify the callback?

File: app/src/customers/customers.controller.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+import { CustomersService } from "./customers.service";
+import { HeaderapikeyAuthGuard } from "../auth/guards/headerapikey-auth.guard";
+
+@Controller('customers')
+export class CustomersController {
+    constructor(
+        private readonly customersService: CustomersService
+    ) {}
+
+    @UseGuards(HeaderapikeyAuthGuard)
```
Comment (erikash) [inline app/src/customers/customers.controller.ts:11]: Love it! ❤️

File: app/src/customers/customers.module.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import { Module } from '@nestjs/common';
+import { ConfigService } from "@nestjs/config";
+import { CustomersService } from "./customers.service";
+import { HeaderApiKeyStrategy } from "../auth/strategies/headerapikey.strategy";
+
+@Module({
+    providers: [
+        CustomersService,
+        ConfigService,
```
Comment (erikash) [inline app/src/customers/customers.module.ts:9]: If you need `ConfigService` import the module don't include it in the providers

File: app/src/customers/customers.module.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+        HeaderApiKeyStrategy,
```
Comment (erikash) [inline app/src/customers/customers.module.ts:10]: Same here - import from `AuthModule`

File: app/src/index.ts [hunk 10]
```diff
@@ -17,6 +17,8 @@ export const handler: Handler<Props> = async (
   context: Context,
   callback: Callback<string>,
 ): Promise<any> => {
+
+  console.log('HANDLER HERE');
```
Comment (erikash) [inline app/src/index.ts:21]: Remove

File: app/src/shared/infra/database/connectingString.ts [hunk 11]
```diff
@@ -22,9 +23,9 @@ const getCloudConnectionStringParams = async (): Promise<IConnectionStringParams
 };
 
 const getLocalConnectionStringParams = async (): Promise<IConnectionStringParams> => {
-  const { HOST, PORT, DB_NAME, USERNAME, PASSWORD } = process.env;
-  if (!HOST || !PORT || !DB_NAME || !USERNAME || !PASSWORD) {
-    const message = `Error, please check ".env" file. (HOST: ${HOST}, PORT: ${PORT}, DB_NAME: ${DB_NAME}, USERNAME: ${USERNAME}, PASSWORD: ${PASSWORD}) keys should exist and have valid values.`;
+  const { HOST, PORT, DB_NAME, USERNAME, PASSWORD, SECRET } = process.env;
+  if (!HOST || !PORT || !DB_NAME || !USERNAME || !PASSWORD || !SECRET) {
```
Comment (erikash) [inline app/src/shared/infra/database/connectingString.ts:27]: Remove SECRET reference

File: app/src/shared/infra/database/connectingString.ts [hunk 12]
```diff
@@ -36,7 +37,8 @@ const getLocalConnectionStringParams = async (): Promise<IConnectionStringParams
     port: Number(PORT),
     dbname: DB_NAME,
     username: USERNAME,
-    password: PASSWORD
+    password: PASSWORD,
+    secret: SECRET,
```
Comment (erikash) [inline app/src/shared/infra/database/connectingString.ts:41]: Remove SECRET reference

File: app/src/shared/infra/database/index.ts [hunk 13]
```diff
@@ -9,12 +9,14 @@ export const initialize = async () => {
...
   const username = connectionStringParams.username;
   const password = connectionStringParams.password;
+  const secret = connectionStringParams.secret;
 
   process.env["HOST"] = host;
   process.env["PORT"] = port;
   process.env["DB_NAME"] = dbname;
   process.env["USERNAME"] = username;
   process.env["PASSWORD"] = password;
+  process.env["SECRET"] = secret;
```
Comment (erikash) [inline app/src/shared/infra/database/index.ts:19]: Remove SECRET reference

File: app/src/users/users.service.ts [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+@Injectable()
+export class UsersService {
+    private readonly users: User[];
+
+    constructor() {
+        this.users = [
+            {
+                userId: 1,
+                username: '+972528321351',
```
Comment (erikash) [inline app/src/users/users.service.ts:13]: Love that the `phone` is the `username`

**Review by erikash — CHANGES_REQUESTED** [review]
Great job! very clean :)
See my minor changes requested...

File: app/src/auth/auth.module.ts (hunk 4 above)
Comment (YuraOkolodko) [reply app/src/auth/auth.module.ts:26]: Config service defined globally so no need import at all.

File: app/src/customers/customers.module.ts (hunk 8 above)
Comment (YuraOkolodko) [reply app/src/customers/customers.module.ts:9]: Same as abowe ConfigService defined globally, no need to import.

File: app/src/customers/customers.module.ts (hunk 9 above)
Comment (YuraOkolodko) [reply app/src/customers/customers.module.ts:10]: Yes config is global - no need to import

**Comment by erikash** [comment]
My point was that you don’t need to define it as a provider in every module
:)

**Review by erikash — APPROVED** [review]
Look great!
Go ahead and merge :)

**Pull request merged**